chips/pc/src/chip.rs

```diff
@@ -0,0 +1,235 @@
+                        self.com2.handle_interrupt();
+                        self.com4.handle_interrupt();
+                    }
+                    36 => {
+                        // Legacy IRQ 4 is shared by COM1 and COM3
+                        self.com1.handle_interrupt();
+                        self.com3.handle_interrupt();
+                    }
+                    _ => unimplemented!("interrupt {num}"),
+                }
+
+                poller.clear_pending(num);
+            }
+        })
+    }
+
+    fn has_pending_interrupts(&self) -> bool {
+        InterruptPoller::access(|poller| poller.next_pending().is_some())
+    }
+
+    #[cfg(target_arch = "x86")]
+    fn sleep(&self) {
+        use x86::bits32::eflags::{self, EFlags};
+
+        // On conventional embedded architectures like ARM and RISC-V, interrupts must be disabled
+        // before going to sleep. But on x86 it is the opposite; we must ensure interrupts are
+        // enabled before issuing the HLT instruction. Otherwise we will never wake up.
+        let eflags = unsafe { eflags::read() };
+        let enabled = eflags.contains(EFlags::FLAGS_IF);
+
+        if enabled {
+            // Interrupts are already enabled, so go ahead and HLT.
+            //
+            // Safety: Assume we are running in ring zero.
+            unsafe {
+                x86::halt();
+            }
+        } else {
+            // We need to re-enable interrupts before HLT-ing. We use inline assembly to guarantee
+            // these instructions are executed back-to-back.
+            //
+            // Safety:
+            //
+            // As above, assume we are running in ring zero.
+            //
+            // Strictly speaking, this could cause to a TOCTOU race condition if `sleep` is called
+            // within an `atomic` block, because interrupt handlers would be executed. Solving this
+            // properly would require deep changes to Tock's `Chip` trait and kernel logic.
+            //
+            // In practice this doesn't seem to be an issue. `sleep` is only ever called once at the
+            // end of the kernel's main loop, and that code does not appear to be vulnerable to the
+            // TOCTOU.
+            unsafe {
+                core::arch::asm!("sti; hlt; cli");
+            }
+        }
+    }
+
+    #[cfg(not(target_arch = "x86"))]
+    fn sleep(&self) {
+        unimplemented!()
+    }
+
+    unsafe fn atomic<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce() -> R,
+    {
+        support::atomic(f)
+    }
+
+    unsafe fn print_state(&self, writer: &mut dyn Write) {
+        let _ = writeln!(writer);
+        let _ = writeln!(writer, "---| PC State |---");
+        let _ = writeln!(writer);
+
+        // todo: print out anything that might be useful
+
+        let _ = writeln!(writer, "(placeholder)");
+    }
+}
+
+/// Component helper for constructing a [`Pc`] chip instance.
```

Comment:
Should this be here or in the `components` folder? The components system was design to help improve code readability when instantiating, but the use of this component somehow seems perform code init stuff that is unavoidable.

Reply:
Good callout, this might merit some discussion...

Internally, our team has been using the components system as a means of _encapsulation_, allowing us to statically ensure that certain init steps are always performed.

We typically make the drivers' fields and constructors totally private, and co-locate the component objects within the same .rs files. This way, clients are forced to use the component system, which in turn ensures that the proper init logic is followed.

I feel it's important to have _something_ that helps with encapsulating init sequences like this, because it helps reduce the complexity of board crates and avoids mistakes or even safety hazards. However I will totally concede that the components system might not be the best fit for this. Maybe `new()` would be a better place?

Reply:
I agree that the init sequence should be coupled with the drivers. I think that the reason for which the components are not in the capsules is the use of `unsafe`. In this case, as this is the `chip` crate, it works.

For the time being, I think using the `new()` function would be a better idea so that it stays consistent with the current implementation of other chips. Would that work for this?

Reply:
I took a closer look at moving the init logic into `new()` instead of using a component. That will work for the chip itself, but there is a slight wrinkle with the serial driver.

The serial driver uses a `DeferredCall` internally to help with cancellation. Calling `DeferredCallClient::register()` requires a `&'static self`, which is not possible inside a conventional `new()` constructor.

Are there examples of chip peripheral drivers elsewhere in Tock that use `DeferredCall` internally? I'd love to take a look at how their initialization is performed so I can follow their pattern here.

Reply:
Hi @reynoldsbd , you're right -- registering Deferred Calls is something we usually do in a component because it cannot be done in `new()`. The apollo3 chip does something similar where most logic happens in new, and then stuff that requires `&'static self` happens in an `init()` function that is just called from boards that use the chip.

Our general policy is not to include `unsafe` in `chips/` except for creating `StaticRef`s for memory-mapped registers (this has not been universally followed but is our goal).

I actually think that components are the right solution for this, and better than what we do for the apollo3 today. But I would prefer the component definition to live in the boards/ crate where other components live. This obviously has the downside that you cannot force consumers to use the component via the use of private fields, but has the upside that unsafe code is contained within `boards/` as Tock would prefer, and that downstream users of the crate can modify initialization of these components as needed from their board file without being forced to modify the chip crate.